llvm/lib/SYCLLowerIR/SYCLJointMatrixTransform.cpp

```diff
@@ -22,6 +22,73 @@ namespace {
 static constexpr char ACCESS_CHAIN[] = "_Z19__spirv_AccessChain";
 static constexpr char MATRIX_TYPE[] = "spirv.CooperativeMatrixKHR";
 
+Type *getInnermostType(Type *Ty) {
+  while (auto *ArrayTy = dyn_cast<ArrayType>(Ty))
+    Ty = ArrayTy->getElementType();
+  return Ty;
+}
+
+Type *replaceInnermostType(Type *Ty, Type *NewInnermostTy) {
+  if (auto *ArrayTy = dyn_cast<ArrayType>(Ty))
+    return ArrayType::get(
+        replaceInnermostType(ArrayTy->getElementType(), NewInnermostTy),
+        ArrayTy->getNumElements());
+  return NewInnermostTy;
+}
+
+// This function is a copy of llvm::stripPointerCastsAndOffsets,
+// simplified and modified to strip non-zero GEP indices as well and also
+// find nearest GEP instruction.
+Value *stripPointerCastsAndOffsets(Value *V, bool StopOnGEP = false) {
+  if (!V->getType()->isPointerTy())
+    return V;
+
+  // Even though we don't look through PHI nodes, we could be called on an
+  // instruction in an unreachable block, which may be on a cycle.
+  SmallPtrSet<Value *, 4> Visited;
+
+  Visited.insert(V);
+  do {
+    if (auto *GEP = dyn_cast<GEPOperator>(V)) {
+      if (StopOnGEP && isa<GetElementPtrInst>(GEP))
+        return V;
+      V = GEP->getPointerOperand();
+    } else if (Operator::getOpcode(V) == Instruction::BitCast) {
+      Value *NewV = cast<Operator>(V)->getOperand(0);
+      if (!NewV->getType()->isPointerTy())
+        return V;
+      V = NewV;
+    } else if (Operator::getOpcode(V) == Instruction::AddrSpaceCast) {
+      V = cast<Operator>(V)->getOperand(0);
+    } else {
+      if (auto *Call = dyn_cast<CallBase>(V)) {
+        if (Value *RV = Call->getReturnedArgOperand()) {
+          V = RV;
+          continue;
+        }
+      }
+      return V;
+    }
+    assert(V->getType()->isPointerTy() && "Unexpected operand type!");
+  } while (Visited.insert(V).second);
+
+  return V;
+}
+
+TargetExtType *extractMatrixType(StructType *WrapperMatrixTy) {
+  if (!WrapperMatrixTy)
+    return nullptr;
+  TargetExtType *MatrixTy =
+      dyn_cast<TargetExtType>(WrapperMatrixTy->getElementType(0));
+
+  if (!MatrixTy)
+    return nullptr;
+  StringRef Name = MatrixTy->getName();
+  if (Name != MATRIX_TYPE)
```

Comment:
nit: can we move the `getName()` call into the `if` since we never use the `Name` variable again and IMO its clear whats going on?

Reply:
done